pkg/generators/forwarding/fluentd/kafka_conf.go

```diff
@@ -0,0 +1,74 @@
+package fluentd
+
+import (
+	"net/url"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultKafkaTopic = "topic"
+
+//Brokers returns the list of broker endpoints of a kafka cluster.
+//The list represents only the initial set used by the collector's kafka client for the
+//first connention only. The collector's kafka client fetches constantly an updated list
+//from kafka. These updates are not reconciled back to the collector configuration.
+//The list of brokers are populated from the Kafka OutputSpec `Brokers` field, a list of
+//valid URLs. If none provided the target URL from the OutputSpec is used as fallback.
+//Finally, if neither approach works the current collector process will be terminated.
+func (conf *outputLabelConf) Brokers() string {
+	parseBroker := func(b string) string {
+		url, err := url.Parse(b)
+		if err != nil {
+			log.Errorf("Failed to parse Kafka broker from output spec %q: %s", b, err)
```

Comment:
We should return errors back to the forwarding code so that informative status conditions can be set for the Output.
This is the first output type that generates errors so the code may not be well factored for this. Shout if you want me to help with designing a good error handling pattern for this.

Reply:
Good catch to re-evaluate this point with an extra test case (See [here]()). Actually we don't need to factor anything out because golang's `text/template` catches every error and will pass it back to the `template.Execute(...)` point of the caller. However, the `log.Fatal` below did not trigger this but it's possible with a `panic`.

Furthermore, regarding the line of code you point to, we should discuss how to handle informative errors. In the highlighted case the helper `parserBroker` function is applied to every broker. We can have valid brokers but also non valid ones on the list. Should we return a error if any of the brokers is broken or just log the error and proceed with a partial list?

Reply:
log.Fatal does os.Exit which is unstoppable :) Panic is OK, we should add a comment somewhere explaining how we use it,  so we don't get a newbie (or forgetful oldie) cut-pasting a panic into a place where it will cause a crash.
 
Outputs are evaluated at reconcile, marked with Conditions, and excluded if they don't have condition Ready=true.  So I think this code (and output-type-specific code in general) should return a Conditions object that can be applied or merged into the Output conditions.

The returned conditions might be like:
* [Ready=true,Degraded=false] - everything works
* [Ready=true,Degraded=true] - have some good and some bad brokers
* [Ready=false,Degraded=false] - total bust

Degrated=true and Ready=False conditions need a Reason and Message. I've been using a small, fixed set of general reasons and putting details (broker names etc.) in the Message. The Reasons are part of the API (Enum-like) so we should be careful with expanding that set. 

I've been composing compound messages when I have multiple things to report, e.g. a Pipeline Message might be something like "not all outputs ready: unready: [foo,bar,baz], degraded:[fred,bill,bob]" 

If we return Conditions they would be merged into the overall status of the Output.